core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java

```diff
@@ -858,4 +858,47 @@ protected MultiTermVectorsResponse executeMultiTermVectors(MultiTermVectorsReque
         throw new UnsupportedOperationException("this test can't handle MultiTermVector requests");
     }
 
+    public final void testParseFromJSON() throws IOException {
+        String json = createParseableQueryJson();
+        QB query = (QB) parseQuery(json);
+        assertQueryParsedFromJson(json, query);
+
+        // now assert that we actually generate the same JSON
+        XContentBuilder builder = XContentFactory.jsonBuilder().prettyPrint();
+        query.toXContent(builder, ToXContent.EMPTY_PARAMS);
+        System.out.println(builder.string());
+        logger.info(msg(json, builder.string()));
```

Comment:
Another idea about the logging here. While this helps at debugging, I think we should only log the query diff when whitespace-normalized queries are not equal (when the following assertion trips), maybe in a finally block?